main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -0,0 +1,127 @@
+import ca.uwaterloo.flix.util.Validation.{ToFailure, ToSuccess}
+
+object Instances extends Phase[TypedAst.Root, TypedAst.Root] {
+
+  /**
+    * Validates instances in the given AST root.
+    */
+  override def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = flix.phase("Instances") {
+    for {
+      _ <- visitInstances(root)
+    } yield root
+  }
+
+
+  /**
+    * Validates all instances in the given AST root.
+    *
+    * Returns [[Err]] if a definition fails to type check.
+    */
+  private def visitInstances(root: TypedAst.Root)(implicit flix: Flix): Validation[Unit, TypeError] = {
+
+    /**
+      * Checks for overlap of instance types, assuming the instances are of the same class.
+      */
+    def checkOverlap(inst1: TypedAst.Instance, inst2: TypedAst.Instance)(implicit flix: Flix): Validation[Unit, TypeError] = {
+      Unification.unifyTypes(inst1.tpe, inst2.tpe) match {
+        case Ok(_) =>
+          Validation.Failure(LazyList(
+            TypeError.OverlappingInstances(inst1.loc, inst2.loc),
+            TypeError.OverlappingInstances(inst2.loc, inst1.loc)
+          ))
+        case Err(_) => ().toSuccess
+      }
+    }
+
+    /**
+      * Checks that every signature in `clazz` is implemented in `inst`, and that `inst` does not have any extraneous definitions.
+      */
+    def checkSigMatch(inst: TypedAst.Instance)(implicit flix: Flix): Validation[Unit, TypeError] = {
+      val clazz = root.classes(inst.sym)
+
+      // Step 1: check that each signature has an implementation.
+      val sigMatchVal = checkEach(clazz.signatures) {
+        sig =>
+          inst.defs.find(_.sym.name == sig.sym.name) match {
+            // Case 1: there is no definition with the same name
+            case None => TypeError.MissingImplementation(sig.sym, inst.loc).toFailure
+            case Some(defn) =>
+              val expectedScheme = Scheme.partiallyInstantiate(sig.sc, clazz.tparam.tpe, inst.tpe)
+              if (Scheme.equal(expectedScheme, defn.declaredScheme, root.classEnv)) {
+                // Case 2.1: the schemes match. Success!
+                ().toSuccess
+              } else {
+                // Case 2.2: the schemes do not match
+                TypeError.MismatchedSignatures(defn.loc, expectedScheme, defn.declaredScheme).toFailure
+              }
+          }
+      }
+      // Step 2: check that there are no extra definitions
+      sigMatchVal.flatMap {
+        _ =>
+          checkEach(inst.defs) {
+            defn =>
+              clazz.signatures.find(_.sym.name == defn.sym.name) match {
+                case None => TypeError.ExtraneousDefinition(defn.sym, defn.loc).toFailure
+                case _ => ().toSuccess
+              }
+          }
+      }
+    }
+
+
+    /**
+      * Reassembles a set of instances of the same class.
+      */
+    def checkInstancesOfClass(insts0: Set[TypedAst.Instance]): Validation[Unit, TypeError] = {
```

Comment:
Can you remind me again of why this is a set and not just a list? What is the notion of equality on TypedAst.Instance? I am almost 100% this should be a list because it is very rare to have sets of anything in the compiler.

Reply:
In the AST, instances are stored as a `MultiMap[ClassSym, Instance]`, which is a wrapper for `Map[ClassSym, Set[Instance]]`. The signature of this function only uses set so I don't have to convert on the outside.

I suppose we could change MultiMap to use ListSet (or maybe just List) to preserve determinism better?